engines/bmc.cpp

```diff
@@ -42,11 +42,16 @@ void Bmc::initialize()
 
 ProverResult Bmc::check_until(int k)
 {
+  // keep the solver instance clean by checking in a different context
+  solver_->push();
   for (int i = 0; i <= k; ++i) {
     if (!step(i)) {
```

Comment:
step asserts unrolled version of the transition relation to the solver. With this change, those assertions will be removed. This is a problem with the current implementation of the step function because it has a check for reached_k (https://github.com/upscale-project/cosa2/blob/49ff98337389ca88a6a478857b5a59704ff7b3de/engines/bmc.cpp#L60).

Reply:
I think that other engines will have the same problem

Reply:
Are you sure? All of the `step` calls are within the same context. `solver_->push()` is called before the calls to `step` and it's not popped until just before a result is returned (after all the `step` calls). That shouldn't be a problem right?